src/OpenApi/src/Extensions/JsonObjectSchemaExtensions.cs

```diff
@@ -0,0 +1,279 @@
+            }
+            if (attribute is MinLengthAttribute minLengthAttribute)
+            {
+                schema["minLength"] = minLengthAttribute.Length;
+            }
+            if (attribute is StringLengthAttribute stringLengthAttribute)
+            {
+                schema["minimum"] = stringLengthAttribute.MinimumLength;
+                schema["maximum"] = stringLengthAttribute.MaximumLength;
+            }
+        }
+    }
+
+    /// <summary>
+    /// Applies the primitive types and formats to the schema based on the type.
+    /// </summary>
+    /// <remarks>
+    /// OpenAPI v3 requires support for the format keyword in generated types. Because the
+    /// underlying schema generator does not support this, we need to manually apply the
+    /// supported formats to the schemas associated with the generated type.
+    ///
+    /// Whereas JsonSchema represents nullable types via `type: ["string", "null"]`, OpenAPI
+    /// v3 exposes a nullable property on the schema. This method will set the nullable property
+    /// based on whether the underlying schema generator returned an array type containing "null" to
+    /// represent a nullable type or if the type was denoted as nullable from our lookup cache.
+    ///
+    /// Note that this method targets <see cref="JsonObject"/> and not <see cref="OpenApiSchema"/> because
+    /// it is is designed to be invoked via the `OnGenerated` callback in the underlying schema generator as
+    /// opposed to after the generated schemas have been mapped to OpenAPI schemas.
+    /// </remarks>
+    /// <param name="schema">The <see cref="JsonObject"/> produced by the underlying schema generator.</param>
+    /// <param name="type">The <see cref="Type"/> associated with the <see paramref="schema"/>.</param>
+    internal static void ApplyPrimitiveTypesAndFormats(this JsonObject schema, Type type)
+    {
+        if (_simpleTypeToOpenApiSchema.TryGetValue(type, out var openApiSchema))
+        {
+            schema["nullable"] = openApiSchema.Nullable || (schema["type"] is JsonArray schemaType
+                && schemaType.GetValues<string>().Contains("null"));
+            schema["type"] = openApiSchema.Type;
+            schema["format"] = openApiSchema.Format;
+        }
+    }
+
+    /// <summary>
+    /// Applies route constraints to the target schema.
+    /// </summary>
+    /// <param name="schema">The <see cref="JsonObject"/> produced by the underlying schema generator.</param>
+    /// <param name="constraints">The list of <see cref="IRouteConstraint"/>s associated with the route parameter.</param>
+    internal static void ApplyRouteConstraints(this JsonObject schema, IEnumerable<IRouteConstraint> constraints)
+    {
+        foreach (var constraint in constraints)
+        {
+            if (constraint is MinRouteConstraint minRouteConstraint)
+            {
+                schema["minimum"] = minRouteConstraint.Min;
```

Comment:
Looks like last one wins if there are multiple contraints and/or attributes?

Reply:
Great question!

It appears that the semantics for route constraints is that the "first-constraint-wins" so we'd actually have to reverse the order in which these are applied.

As for attributes, it looks like _both_ validations must be met. So if you have something like:

```
[FromRoute]
[MinLength(5)]
[StringLength(10, MinimumLength = 6)]
public string? Name { get; set; }
```

Then the realized minimum length is 6 (the larger of the values) regardless of which order the attributes appear in. To address this, we'd have to be cleverer and track the actual value to be evaluated against. I'll see if we can achieve this without making the code too complicated.

There's also the fact that route constraints trump validation attributes because they are evaluated first in the request pipeline. We already handle this well because route constraints are applied last.

I'll add a few more test cases to cover this area.